### PR TITLE
Refactor multi attribute login user resolving logic

### PR DESCRIPTION
**Pull request by sandushi**
### Proposed changes in this pull request
- Change the multi attribute login user solving logic to resolve the user with username by default when there is no user found from the regex matching, 
- So, the user is checking from the username claim.
- Improve the logic to search from the username when there is no users found from the regex patterns matching.

### CODE REVIEW

File: components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java [hunk 1]
```diff
@@ -138,6 +127,15 @@ private void resolveDistinctUsersForClaims(String loginAttribute, List<String> a
             }
         }
 
+        // Check the users from username by default if there are users found from claims and regex patterns.
```
Comment (lashinijay) [inline components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java:130]:
```suggestion
        // Check the users from username by default if there are no users found from claims and regex patterns.
```

**Review by inthirakumaaran — DISMISSED** [review]

File: components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java [hunk 2]
```diff
@@ -138,6 +127,15 @@ private void resolveDistinctUsersForClaims(String loginAttribute, List<String> a
             }
         }
 
+        // Check the users from username by default if there are no users found from claims and regex patterns.
+        if (distinctUsers.size() == 0) {
+            if (allowedAttributes.contains(UserCoreClaimConstants.USERNAME_CLAIM_URI)) {
+                List<User> userList = userStoreManager.getUserListWithID(UserCoreClaimConstants.USERNAME_CLAIM_URI,
+                        loginAttribute, null);
```
Comment (sandushi) [inline components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java:134]: check that the same user is adding to the list

**Review by lashinijay — APPROVED** [review]

File: components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java (hunk 2 above)
Comment (sandushi) [reply components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java]: Resolved from https://github.com/wso2-extensions/identity-governance/pull/743/commits/5cff3e814be7727922bf3bc29413cc3c71fea72d

File: components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java [hunk 3]
```diff
@@ -138,6 +128,22 @@ private void resolveDistinctUsersForClaims(String loginAttribute, List<String> a
             }
         }
 
+        // Check the users from username by default if there is no regex for username claim.
+        Claim usernameClaim = claimManager.getClaim(UserCoreClaimConstants.USERNAME_CLAIM_URI);
+        if (allowedAttributes.contains(UserCoreClaimConstants.USERNAME_CLAIM_URI)
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java:133]: Do we need to check for username claim in the allowed list i.e any future possibility to remove username

Comment (sandushi) [reply components/org.wso2.carbon.identity.multi.attribute.login/org.wso2.carbon.identity.multi.attribute.login.resolver.regex/src/main/java/org/wso2/carbon/identity/multi/attribute/login/resolver/regex/RegexResolver.java:133]: With the API we can update the attribute list without the username claim . So it's better to check that the username is existing in the list.

**Review by inthirakumaaran — APPROVED** [review]

**Pull request merged**